django/db/models/base.py

```diff
@@ -1108,6 +1108,26 @@ def _prepare_related_fields_for_save(self, operation_name, fields=None):
                 ):
                     field.delete_cached_value(self)
 
+        # GenericForeignKey fields are private
+        for field in self._meta.private_fields:
+            if (
+                field.is_relation
+                and field.is_cached(self)
+                and hasattr(field, "fk_field")
+            ):
+                obj = field.get_cached_value(self, default=None)
+                if not obj:
+                    continue
+                if obj.pk is None:
+                    raise ValueError(
+                        "%s() prohibited to prevent data loss due to unsaved "
+                        "related object '%s'." % (operation_name, field.name)
+                    )
+                elif getattr(self, field.fk_field) is None:
+                    # Use pk from related object if it has been saved after
+                    # an assignment.
+                    setattr(self, field.fk_field, obj.pk)
```

Comment:
This is a separate issue and we should move it to a separate commit (together with `test_save_generic_foreign_key_after_parent`) e.g.
> _"Refs #28147 -- Fixed loss of assigned parent when saving GenericForeignKey after parent."_
Maybe even a separate PR to make it easier for review.

Reply:
We should also check if this doesn't introduce a regression as 519016e5f25d7c0a040015724f9920581551cab0 (fixed in 4122d9d3f1983eea612f236e941d937bd8589a0d).

Reply:
I've removed and I'll raise a separate PR once we're happy with this one :+1: